### PR TITLE
improve password handling

### PR DESCRIPTION
**Pull request by craigh**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| New feature?      | yes
| BC breaks?        | yes
| Deprecations?     | yes
| Fixed tickets     | #2842
| Refs tickets      | -
| License           | MIT
| Changelog updated | yes

## Description

improve password handling

### CODE REVIEW

File: CHANGELOG-3.0.md [hunk 1]
```diff
@@ -76,7 +76,9 @@
     - The corresponding Twig function is similarly renamed.
   - `Zikula\ExtensionsModule\Event\ModuleStateEvent` is renamed to `Zikula\ExtensionsModule\Event\ExtensionStateEvent`.
     - Its methods also renamed: `getModule` -> `getExtension` and `getModInfo` -> `getInfo`.
-  - All the Events in `Zikula\ExtensionsModule\ExtensionEvents` are changed - both the name and the ConstantName. 
+  - All the Events in `Zikula\ExtensionsModule\ExtensionEvents` are changed - both the name and the ConstantName.
+  - `Zikula\ZAuthModule\Api\PasswordApi` & `Zikula\ZAuthModule\Api\ApiInterface\PasswordApiInterface` are deprecated and will be removed in Core-4.0.0
```
Comment (Guite) [inline CHANGELOG-3.0.md]: Why not remove now?

File: CHANGELOG-3.0.md [hunk 2]
```diff
@@ -198,6 +200,7 @@
   - Start page can now be defined much easier (a dropdown allows to choose a route/controller combination) (#3955).
   - Start page arguments can now be defined more flexible (GET parameters and request attributes) (#3955).
   - Start page can now be configured for each available language (#3955).
+  - Passwords in the ZAuth module are now always hashed with the the most up-to-date algorithm available (via Symfony security component) and automatically updated on login (#2842).
```
Comment (Guite) [inline CHANGELOG-3.0.md]: Nice :+1:

File: CHANGELOG-3.0.md (hunk 1 above)
Comment (craigh) [reply CHANGELOG-3.0.md]: because we wouldn't be able to upgrade passwords from existing userbase

**Comment by Guite** [comment]
Please check Travis before merging.

File: CHANGELOG-3.0.md (hunk 1 above)
Comment (craigh) [reply CHANGELOG-3.0.md]: removed what I could.

**Comment by craigh** [comment]
@Guite please review again

File: src/system/ZAuthModule/Controller/UserAdministrationController.php [hunk 3]
```diff
@@ -462,4 +465,93 @@ public function togglePasswordChangeAction(Request $request, UserEntity $user)
...
+                $count = 0;
+                foreach ($users as $user) {
+                    $user->setAttribute(ZAuthConstant::REQUIRE_PASSWORD_CHANGE_KEY, true);
+                    $count++;
+                    if ($count % 20 === 0) {
+                        $this->getDoctrine()->getManager()->flush(); // flush manager every 20 reps
+                    }
+                }
+                $this->getDoctrine()->getManager()->flush(); // flush remaining
+                /** @Desc("{count, plural,\n  one   {# user was processed!}\n  other {# users were processed!}\n}") */
```
Comment (Guite) [inline src/system/ZAuthModule/Controller/UserAdministrationController.php]: Missing import: `use Translation\Extractor\Annotation\Desc;`

File: src/system/ZAuthModule/Entity/Repository/AuthenticationMappingRepository.php [hunk 4]
```diff
@@ -20,6 +20,7 @@
 use Zikula\Bundle\CoreBundle\Doctrine\WhereFromFilterTrait;
 use Zikula\ZAuthModule\Entity\AuthenticationMappingEntity;
 use Zikula\ZAuthModule\Entity\RepositoryInterface\AuthenticationMappingRepositoryInterface;
+use function Doctrine\ORM\QueryBuilder;
```
Comment (Guite) [inline src/system/ZAuthModule/Entity/Repository/AuthenticationMappingRepository.php]: Is this import actually needed?

File: src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        $this->groupRepository = $groupRepository;
+    }
+
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('group', ChoiceType::class, [
+                'label' => 'Group to modify',
+                'required' => true,
+                'choices' => $this->getChoices(),
```
Comment (Guite) [inline src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php]: This needs to be `'choices' => /** @Ignore */$this->getChoices(),` with `use Translation\Extractor\Annotation\Ignore;`

File: src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                'help' => 'Old passwords are pre-Core3 passwords which are less secure. <strong>The current user will not be affected</strong>.',
+                'help_html' => true
+            ])
+            ->add('submit', SubmitType::class, [
+                'label' => 'Submit',
+                'icon' => 'fas fa-check',
```
Comment (Guite) [inline src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php]: Remove `fas `

File: src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                'attr' => [
+                    'class' => 'btn-success'
+                ]
+            ])
+            ->add('cancel', SubmitType::class, [
+                'label' => 'Cancel',
+                'icon' => 'fas fa-times'
```
Comment (Guite) [inline src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php]: Remove `fas `

File: src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            ])
+        ;
+    }
+
+    private function getChoices(): array
+    {
+        $choices = [
+            'Users with old passwords (recommended)' => 'old',
+            'All users' => 'all'
+        ];
```
Comment (Guite) [inline src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php]: This needs to be:
```php
        $choices = [
            /** @Translate */'Users with old passwords (recommended)' => 'old',
            /** @Translate */'All users' => 'all'
        ];
```
with `use Translation\Extractor\Annotation\Translate;`

File: src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        /** @var \Zikula\GroupsModule\Entity\GroupEntity[] $groups */
+        $groups = $this->groupRepository->getGroups();
+        foreach ($groups as $group) {
+            $choices[$group->getName() . ' group'] = $group->getGid();
```
Comment (Guite) [inline src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php]: I suggest to remove ` . ' group'`

File: src/system/ZAuthModule/Resources/public/js/README.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
+Included Javascript vendors
+===========================
+
+pwstrength-bootstrap.min.js
+---------------------------
+https://github.com/ablanco/jquery.pwstrength.bootstrap
+
+License: MIT
```
Comment (Guite) [inline src/system/ZAuthModule/Resources/public/js/README.md]: Please replace with:
```markdown
# Included Javascript vendors

## pwstrength-bootstrap.min.js

https://github.com/ablanco/jquery.pwstrength.bootstrap

License: MIT
```

File: src/system/ZAuthModule/Resources/public/js/Zikula.ZAuth.PWStrength.Init.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+        options.common.usernameField = '#' + $(this).data('uname-id');
+      }
+      options.ui = {
+        showVerdictsInsideProgressBar: true,
+        progressExtraCssClasses: "mt-2"
+      };
+      $(this).pwstrength(options);
+    });
+  });
+})(jQuery);
```
Comment (Guite) [inline src/system/ZAuthModule/Resources/public/js/Zikula.ZAuth.PWStrength.Init.js]: Indentation should be 4 spaces.

File: src/system/ZAuthModule/Resources/public/js/Zikula.ZAuth.PasswordGenerator.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+      passwordInput.val(createPassword());
+    });
+    $('#copy-button').on('click', function() {
+      passwordInput.focus();
+      passwordInput.select();
+      document.execCommand('copy');
+      passwordModal.modal('hide')
+    });
+  });
+})(jQuery);
```
Comment (Guite) [inline src/system/ZAuthModule/Resources/public/js/Zikula.ZAuth.PasswordGenerator.js]: Indentation should be 4 spaces.

File: src/system/ZAuthModule/Entity/Repository/AuthenticationMappingRepository.php (hunk 4 above)
Comment (Guite) [reply src/system/ZAuthModule/Entity/Repository/AuthenticationMappingRepository.php]: Seems not

File: src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php (hunk 9 above)
Comment (craigh) [reply src/system/ZAuthModule/Form/Type/BatchForcePasswordChangeType.php]: group is needed

**Pull request merged**